src/program.ts

```diff
@@ -3079,7 +3149,26 @@ export class Function extends TypedElement {
   lookup(name: string): Element | null {
     var locals = this.localsByName;
     if (locals.has(name)) return assert(locals.get(name));
-    return this.parent.lookup(name);
+    var parentResult = this.parent.lookup(name);
+    if(parentResult && this.parent.kind == ElementKind.FUNCTION) {
+      var parentFunction = <Function>this.parent;
+      if(parentFunction.closedLocals.size > 0) { //TODO allow nested closure definitions
+        this.program.error(
+          DiagnosticCode.Not_implemented,
+          this.identifierNode.range, this.identifierNode.text
+        );
+        return null;
+      }
+      if(parentResult.kind == ElementKind.LOCAL) {
+        let local = changetype<Local>(parentResult)
+        if(this.closedLocals.has(local.name)) return assert(this.closedLocals.get(local.name))
+        var closedLocal = local.close(this.nextGlobalClosureOffset);
+        this.nextGlobalClosureOffset += local.type.byteSize;
```

Comment:
the `nextGlobalClosureOffset` hack is here because we haven't generated the class yet at this point- so based on an assumption on how the fields will be ordered and packed, we determine the offset of each local. 

This could be replaced if we had a mechanism for determining where we had closures before we began compiling

Reply:
Also looks like this should take alignment into account to prevent unaligned loads and stores.

Reply:
Added in some alignment logic that we use when assigning memory offsets for fields- is this what you meant?

Reply:
Yes, looks good this way. That makes sure that the loads are fast (unaligned ones are rather unpredictable, often much slower).